runtime/src/snapshot_utils.rs

```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+            archive_format,
+            parallel_divisions,
+        )?;
+        (
+            Some(incremental_snapshot_unpack_dir),
+            Some(incremental_snapshot_unpacked_snapshots_dir),
+            Some(incremental_snapshot_unpacked_append_vec_map),
+            Some(incremental_snapshot_version),
+            Some(incremental_snapshot_measure_untar),
+        )
+    } else {
+        (None, None, None, None, None)
+    };
+
+    let mut unpacked_append_vec_map = full_snapshot_unpacked_append_vec_map;
+    unpacked_append_vec_map.extend(
+        incremental_snapshot_unpacked_append_vec_map
+            .unwrap_or_default()
+            .into_iter(),
+    );
 
+    let mut measure_rebuild = Measure::start("rebuild bank from snapshots");
     let bank = rebuild_bank_from_snapshots(
-        snapshot_version.trim(),
+        full_snapshot_version.trim(),
+        incremental_snapshot_version
+            .as_ref()
+            .map(|version| version.trim()),
         frozen_account_pubkeys,
-        &unpacked_snapshots_dir,
+        &full_snapshot_unpacked_snapshots_dir,
+        incremental_snapshot_unpacked_snapshots_dir.as_ref(),
         account_paths,
         unpacked_append_vec_map,
         genesis_config,
         debug_keys,
         additional_builtins,
-        account_indexes,
+        account_secondary_indexes,
         accounts_db_caching_enabled,
         limit_load_slot_count_from_snapshot,
         shrink_ratio,
     )?;
-    measure.stop();
+    measure_rebuild.stop();
+    info!("{}", measure_rebuild);
 
-    let mut verify = Measure::start("verify");
+    let mut measure_verify = Measure::start("verify");
     if !bank.verify_snapshot_bank(test_hash_calculation)
         && limit_load_slot_count_from_snapshot.is_none()
     {
         panic!("Snapshot bank for slot {} failed to verify", bank.slot());
     }
-    verify.stop();
+    measure_verify.stop();
+
     let timings = BankFromArchiveTimings {
-        rebuild_bank_from_snapshots_us: measure.as_us(),
-        untar_us: untar.as_us(),
-        verify_snapshot_bank_us: verify.as_us(),
+        rebuild_bank_from_snapshots_us: measure_rebuild.as_us(),
+        untar_us: full_snapshot_measure_untar.as_us()
+            + incremental_snapshot_measure_untar.map_or(0, |measure| measure.as_us()),
+        verify_snapshot_bank_us: measure_verify.as_us(),
     };
-
     Ok((bank, timings))
 }
 
+/// Perform the common tasks when unarchiving a snapshot.  Handles creating the temporary
+/// directories, untaring, reading the version file, and then returning those fields plus the
+/// unpacked append vec map.
+fn do_snapshot_unarchive_preparation<P, Q>(
+    snapshot_path: P,
+    unpacked_snapshots_dir_prefix: &'static str,
+    snapshot_archive_path: Q,
+    measure_name: &'static str,
+    account_paths: &[PathBuf],
+    archive_format: ArchiveFormat,
+    parallel_divisions: usize,
+) -> Result<(TempDir, PathBuf, UnpackedAppendVecMap, String, Measure)>
```

Comment:
This is a rather large tuple that could deserve a struct. Might be good to also include the struct described here: https://github.com/solana-labs/solana/pull/18504/files#r668273623   as a sub-struct of that struct

Reply:
I agree! Will do.

Reply:
Done!